### ISSUE TITLE
[Feature Request] Use Kotlin

### ISSUE THREAD

**Issue by oleersoy** [Hipparchus-Math/hipparchus#26]
Kotlin has a lot of slick features, like eliminating the need for builders, etc. through named arguments, that I think would be useful for Hipparchus.  Also you can compile to both Javascript and various versions of JVMs.  See:

https://stackoverflow.com/questions/46858270/does-there-exist-a-babel-like-compiler-for-java
https://stackoverflow.com/questions/46892929/are-number-operations-using-kotlin-as-fast-as-the-equivalent-with-java-primitive

General info:
https://medium.com/@magnus.chatt/why-you-should-totally-switch-to-kotlin-c7bbde9e10d5

**Comment by maisonobe** [comment]
I didn't know about Kotlin.
The one thing I would love is its operator overloading feature. It would really help using our RealField implementations (I do use DerivativeStructure a *lot*). It seems powerful (it even has things like +=), but still seem to lack infix functions without parameters (i.e. you can't replace x.cos() with cos(x)).

**Comment by oleersoy** [comment]
Yeah I just came across it a week ago catching up on some Reactive programming with Spring.  It looks like a lot of fun to program with.  All the latest Java goodies, but far less verbose, and great support from IntelliJIdea (The creators).  And as I mentioned you can choose to compile to any Java runtime, so kiss the "Should we move to Java ... discussion" goodbye.  Should help a lot with the fluid API stuff too, as constructors support named arguments.

**Comment by oleersoy** [comment]
Compiling to Javascript could add a lot of value as well.  For example searching for linear programming on https://www.npmjs.com/ leads to packages / modules like:

https://www.npmjs.com/package/javascript-lp-solver

This has very good documentation and ( I just found it) looks like its very easy to use and play around with.  So if Hipparchus compiles to javascript and releases to NPM it gets a whole new set of testers / users.

**Comment by oleersoy** [comment]
Incidentally I created a hipparchus-math organization on NPM just so that no one else grabs it.

**Comment by wardev** [comment]
At this point I'm not for or against the proposal, but I think there are a couple more points to consider. 

Using Kotlin would add a runtime dependency for Hipparchus because Kotlin comes with its own standard library. This would be Hipparchus' first external dependency.

As far as I can tell the only IDE with good support for Kotlin is IDEA, which means all Hipparchus developers would be forced to use IDEA. I don't mind because I already use it, but we should make sure the other developers are OK with it.

We currently have extensive tooling that works well for Java, i.e. Javadoc, JXR, Findbugs, Jacoco, and Checkstyle.  Are there equivalents for Kotlin? Or do our existing maven plugins support Kotlin?

**Comment by oleersoy** [comment]
Wow - I did not know it created a runtime dependency - thanks for pointing that out.  

There are is an [eclipse plugin for kotlin](https://kotlinlang.org/docs/tutorials/getting-started-eclipse.html) - although it looks fairly new.  Atom also has a [syntax highlighter](https://atom.io/packages/language-kotlin)  (lightweight tasks).  

When I looked up static code analysis for Kotlin I came across [detekt which looks interesting](https://github.com/arturbosch/detekt).  

There's also [Dokka for documentation](https://github.com/Kotlin/dokka).  Once you start getting into the Javascript / Typescript world there are some crazy good tools - like [compodoc for Angularjs](https://github.com/compodoc/compodoc).  Kotlin looks a lot like Typescript so I suspect we'll see a lot more goodness like that arriving in the Kotlin ecosystem.

Kotlin has [ktlint](https://github.com/shyiko/ktlint) for linting.

**Comment by shiMusa** [comment]
Hi all,
I'm using (until now, changing that now) the Apache Math library in Kotlin for my research and really, Kotlin makes so many things so much easier, especially higher-order functions would be awesome for this kind of math (instead of using classes for everything e.g.). Another feature is the very easy to use coroutine system for parallel execution.
In addition to JS, there is also the thought of using Kotlin-Multiplatform such that the math can be run natively.
I don't suggest rewriting the whole library in Kotlin, but some Kotlin wrappers would already be amazing ;)

Hope this project keeps growing! Good luck :)

**Comment by HankG** [comment]
@shiMusa I'm a huge fan of Kotlin and use it on JVM projects when I have the choice. Is there something that the standard Java bridge in Kotlin isn't doing well or you are saying adding some Kotlin specific decorators could make it even better?
